hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataWriteUtils.java

```diff
@@ -84,8 +89,20 @@ public class HoodieMetadataWriteUtils {
    */
   @VisibleForTesting
   public static HoodieWriteConfig createMetadataWriteConfig(
-      HoodieWriteConfig writeConfig, HoodieFailedWritesCleaningPolicy failedWritesCleaningPolicy) {
+      HoodieWriteConfig writeConfig, HoodieFailedWritesCleaningPolicy failedWritesCleaningPolicy,
+      HoodieTableVersion datatableVersion) {
     String tableName = writeConfig.getTableName() + METADATA_TABLE_NAME_SUFFIX;
+    boolean isStreamingWritesToMetadataEnabled = writeConfig.isStreamingWritesToMetadataEnabled(datatableVersion);
+    WriteConcurrencyMode concurrencyMode = isStreamingWritesToMetadataEnabled
+        ? WriteConcurrencyMode.NON_BLOCKING_CONCURRENCY_CONTROL : WriteConcurrencyMode.SINGLE_WRITER;
+    HoodieLockConfig lockConfig = isStreamingWritesToMetadataEnabled
+        ? HoodieLockConfig.newBuilder().withLockProvider(InProcessLockProvider.class)
```

Comment:
Shouldn't the lock config come from the data table? Is it also possible to share the lock provider used with the data table writer? That way we can leverage the re-entrant nature of the locks and avoid deadlock bugs.

Reply:
not all prod grade lock providers have re-entrant support. So, to not complicate things, we have kept it simple. we don't have any re-entrancy anywhere in our code.

Reply:
I think all the lock providers have re-entrant support, which one does not?